src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -0,0 +1,731 @@
+            _origin = origin;
+            _authority = authority;
+            _connection = connection;
+
+            bool altUsedDefaultPort = pool.Kind == HttpConnectionKind.Http && authority.Port == HttpConnectionPool.DefaultHttpPort || pool.Kind == HttpConnectionKind.Https && authority.Port == HttpConnectionPool.DefaultHttpsPort;
+            string altUsedValue = altUsedDefaultPort ? authority.IdnHost : authority.IdnHost + ":" + authority.Port.ToString(Globalization.CultureInfo.InvariantCulture);
+            _altUsedEncodedHeader = QPack.QPackEncoder.EncodeLiteralHeaderFieldWithoutNameReferenceToArray(KnownHeaders.AltUsed.Name, altUsedValue);
+
+            _maximumRequestStreams = _requestStreamsRemaining = connection.GetInitialMaximumBidirectionalStreams();
+
+            // Errors are observed via Abort().
+            _ = SendSettingsAsync();
+
+            // This process is cleaned up when _connection is disposed, and errors are observed via Abort().
+            _ = AcceptStreamsAsync();
+        }
+
+        /// <summary>
+        /// Starts shutting down the <see cref="Http3Connection"/>. Final cleanup will happen when there are no more active requests.
+        /// </summary>
+        public void Dispose()
+        {
+            lock (SyncObj)
+            {
+                if (_lastProcessedStreamId == -1)
+                {
+                    _lastProcessedStreamId = long.MaxValue;
+                    CheckForShutdown();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called when shutting down, this checks for when shutdown is complete (no more active requests) and does actual disposal.
+        /// </summary>
+        /// <remarks>Requires <see cref="SyncObj"/> to be locked.</remarks>
+        private void CheckForShutdown()
+        {
+            Debug.Assert(Monitor.IsEntered(SyncObj));
+            Debug.Assert(ShuttingDown);
+
+            if (_activeRequests.Count != 0)
+            {
+                return;
+            }
+
+            if (_clientControl != null)
+            {
+                _clientControl.Dispose();
+                _clientControl = null;
+            }
+
+            if (_connection != null)
+            {
+                _connection.Close((long)Http3ErrorCode.NoError);
+                _connection.Dispose();
+                _connection = null;
+            }
+        }
+
+        public override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Task waitTask = WaitForAvailableRequestStreamAsync(cancellationToken);
+
+            return waitTask.IsCompletedSuccessfully
+                ? SendWithoutWaitingAsync(request, cancellationToken)
+                : WaitThenSendAsync(waitTask, request, cancellationToken);
+        }
+
+        private async Task<HttpResponseMessage> WaitThenSendAsync(Task taskToWaitFor, HttpRequestMessage request, CancellationToken cancellationToken)
```

Comment:
nit: ValueTask all the things?

Reply:
For now, only the things I expect to finish synchronously.

In the future, depending on the outcome of @stephentoub's `ValueTask` caching patch, maybe `ValueTask` all the things.

Reply:
I think it's reasonable to use `ValueTask<T>` for any of these that are internal/private, as long as we're only ever awaiting them at the call sites.  That will help us to better evaluate the ValueTask opt-in change I made.

Reply:
@stephentoub I'm open to it, but how will it help us evaluate things when there's no baseline to compare to?

Reply:
The comparison I'm referring to is when lots of stuff uses `ValueTask<T>`, and we flip the switch to enable automatic pooling.  If the method completes synchronously, then using `ValueTask<T>` was a good idea, and if it completes asynchronously, then the overhead compared to the allocation/state management/etc. is minimal, so I'm not particularly concerned with a comparison about HTTP/3's implementation with `ValueTask<T>` vs `Task<T>` here.

Reply:
Thanks. I'll leave this for a post-merge update.